demo/src/main/java/com/google/android/fhir/demo/data/TimestampBasedDownloadWorkManagerImpl.kt

```diff
@@ -112,16 +144,63 @@ class TimestampBasedDownloadWorkManagerImpl(
           .map { it.resource }
           .also { extractAndSaveLastUpdateTimestampToFetchFutureUpdates(it) }
       for (item in response.entry) {
-        if (item.resource is PlanDefinition) {
-          bundleCollection +=
-            carePlanManager.getPlanDefinitionDependentResources(item.resource as PlanDefinition)
-          bundleCollection += getCareConfigurationResources()
-        }
+        bundleCollection += processResourceForExtraction(item.resource)
       }
     }
+    println("Bundle size is "+ bundleCollection.size)
     return bundleCollection
   }
 
+  private suspend fun processResourceForExtraction(resource: Resource): Collection<Resource> {
+    when (resource) {
+      is PlanDefinition -> return extractPlanDefinitionDependentResources(resource)
+      is CarePlan -> return extractCarePlanDependentResources(resource)
+      is Encounter -> return addEncounterRelatedResources(resource)
+    }
+    return emptyList()
+  }
+
+  private suspend fun extractPlanDefinitionDependentResources(planDefinition: PlanDefinition): Collection<Resource> {
+    //get all CarePlans for the relevant Patients that are created as a part of execution of this PlanDefinition
+    if (planDefinition.url != null) {
+      addResourceIdsToSearch(ResourceType.CarePlan, "instantiates-canonical", setOf(planDefinition.url))
+    }
+    return carePlanManager.getPlanDefinitionDependentResources(planDefinition) + getCareConfigurationResources()
+  }
+
+  private fun extractCarePlanDependentResources(carePlan: CarePlan): Collection<Resource> {
+    //get all the tasks that have been carried out/ need to be carried out as a part of this CarePlan
+    val taskIds = carePlan.activity.map { getResourceIdFromReference(it.reference) }.toSet()
+    addResourceIdsToSearch(ResourceType.Task, "_id", taskIds)
+    //get all the encounters that were created as a result of carrying out tasks of this CarePlan
+    val encounterIds = carePlan.activity.flatMap { it.outcomeReference.map { or -> getResourceIdFromReference(or) } }.toSet()
+    addResourceIdsToSearch(ResourceType.Encounter, "_id", encounterIds)
+    return emptyList()
+  }
+
+  private fun addEncounterRelatedResources(encounter: Encounter): Collection<Resource> {
+    //get all related observations and conditions for these encounters
+    addResourceIdsToSearch(ResourceType.Observation, "encounter", setOf(encounter.idElement.idPart))
+    addResourceIdsToSearch(ResourceType.Condition, "encounter", setOf(encounter.idElement.idPart))
+    return emptyList()
+  }
+
+  private fun getResourceIdFromReference(reference: Reference): String {
+    val referenceElements = reference.reference.split("/")
+    return referenceElements[1]
+  }
+
+  private fun addResourceIdsToSearch(resourceType: ResourceType, searchParam: String, resourceIds: Set<String>) {
```

Comment:
May reduce code using merge. Something like this. Not tested though.

resourceReferences.merge(resourceType, mutableMapOf(resourceParam to resourceIds)) {
      oldValue, _ ->
      mutableMapOf(resourceParam to oldValue[resourceParam].intersect(resourceIds))
    }